✅(frontend) add  visual regression testing (#1222)
openfun/marsha: docs/dev.md

docker-compose exec app python manage.py test marcha.path.to.module.Class.method
```

### Visual regression testing

Visual regression testing is available through [jest-image-snapshot](https://github.com/americanexpress/jest-image-snapshot)
and [jsdom-screenshot](https://github.com/dferber90/jsdom-screenshot).
Two wrappers have been added in `src/frontend/utils/tests/imageSnapshot.ts`

#### imageSnapshot(width?: number, height?: number)

> Simply takes a screenshot and compares it to the previous one if it exists.

Reviewer: ```suggestion
Takes a snapshot of the current DOM generated in your test and compares it to the previous one if it exists.
```

**Current test states will be kept.**

#### renderImageSnapshot(component: JSX.Element, width?: number, height?: number)

> Renders passed component wrapped with wrapInIntlProvider, wrapInRouter and Grommet.

Reviewer: ```suggestion
Renders passed component wrapped in `wrapInIntlProvider`, `wrapInRouter` and `Grommet`.
```

> Renders passed component wrapped with wrapInIntlProvider, wrapInRouter and Grommet.

Reviewer: ```suggestion
`renderImageSnapshot` is a helper using `imageSnapshot` and renders for you the component passed in parameter. The component will be wrappred in `wrapInIntlProvider`, `wrapInRouter` and `Grommet`.
```

GlobalStyles are added to.

> imageSnapshot is called afterwards.

Reviewer: ```suggestion
```

> **Current test states will be lost.**

Reviewer: This can be a problem If you want to take snapshot at different states of your component.

Author: Indeed, that's why I added `imageSnapshot` which only takes screenshot and doesn't render the component.

Reviewer: ```suggestion
**This helper will destroy your current DOM generated by you test and will replace it by the one generated in this function.**
```

#### renderIconSnapshot(component: JSX.Element)

Calls renderImageSnapshot with with=100 and height=100.

> Particularly helpful for testing SVG components.

Reviewer: Here explain it is an helper too maybe

docker-compose exec app python manage.py test marcha.path.to.module.Class.method
```

### Visual regression testing

Visual regression testing is available through [jest-image-snapshot](https://github.com/americanexpress/jest-image-snapshot)
and [jsdom-screenshot](https://github.com/dferber90/jsdom-screenshot).
Two wrappers have been added in `src/frontend/utils/tests/imageSnapshot.ts`

#### imageSnapshot(width?: number, height?: number)

**Current test states will be kept.**

`imageSnapshot` takes a snapshot of the current DOM generated in your test and compares it to the previous one if it exists.

#### renderImageSnapshot(component: JSX.Element, width?: number, height?: number)

> **This helper will destroy your current DOM generated by your test and will replace it by the one generated in this function.**

Reviewer 2: ```suggestion
**This helper will destroy your current DOM generated by your test and will replace it with the one generated in this function.**
```

`renderImageSnapshot` is a helper using `imageSnapshot` and renders for you the component passed in parameter.

> The component will be wrappred in `wrapInIntlProvider`, `wrapInRouter` and `Grommet`.

Reviewer 3: ```suggestion
The component will be wrapped in `wrapInIntlProvider`, `wrapInRouter` and `Grommet`.
```

> The component will be wrappred in `wrapInIntlProvider`, `wrapInRouter` and `Grommet`.

Reviewer 2: ```suggestion
The component will be wrapped in `wrapInIntlProvider`, `wrapInRouter` and `Grommet`.
```

GlobalStyles are added to.

> This is useful for tests that doesn't render any component.

Reviewer 3: ```suggestion
This is useful for tests that don't render any component.
```

> This is useful for tests that doesn't render any component.

Reviewer 2: ```suggestion
This is useful for tests that don't render any component.
```

#### renderIconSnapshot(component: JSX.Element)

**This helper will destroy your current DOM generated by your test and will replace it by the one generated in this function.**

`renderIconSnapshot` is a helper using `renderImageSnapshot` called with predefined height and width.

This is useful for testing icons components.

#### Usage

Import the wrapper and call imageSnapshot in a component async test to take a snapshot.

```typescript jsx
import { imageSnapshot } from 'utils/tests/imageSnapshot';

[…]

await imageSnapshot(300, 100);
await renderImageSnapshot(<AppsSVG iconColor={'#035ccd'} />, 200, 200)
await renderIconSnapshot(<AppsSVG iconColor={'#035ccd'} />)
```

> When the test is run for the first time, a reference image snapshot is taken, stored on a `__image_snapshots__`directory

Reviewer 3: ```suggestion
When the test is run for the first time, a reference image snapshot is taken, stored on a `__image_snapshots__` directory
```

alongside the test file.

**The reference image snapshot must be committed.**

![image_snapshot_1](images/image_snapshot_1.png)

Please note that multiple image snapshots can be taken inside a test.

> Now, everytime the test is run, a new image snapshot is taken and compared to the previous one.

Reviewer: ```suggestion
Now, everytime the test is run, a new image snapshot is generated and compared to the previous one.
```

> Now, everytime the test is run, a new image snapshot is generated and compared to the previous one.

Reviewer 2: ```suggestion
Now, every time the test is run, a new image snapshot is generated and compared to the previous one.
```

If a difference appears, a new image showing both image snapshots and the differences is generated
and stored in a `__diff_output__` directory inside `/src/frontend`.

![image_snapshot_diff_1](images/image_snapshot_diff_1.png)

The developer is responsible to accept the difference,

> and can update the reference image snapshot with following command:

Reviewer 2: ```suggestion
and can update the reference image snapshot with the following command:
```

The new image snapshot must be committed and be approved like the code in a pull request.
If the pull request is merged, the image snapshot will be the new reference.

> During tests run on Circle CI, differences detected be available in artifacts tab.

Reviewer: Still missing words 😁

Author: :facepalm:
